### PR TITLE
[Pipeline Inference] Merge pp with tp

### PR DESCRIPTION
**Pull request by FoolPlayer**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.
Merge pp with tp

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/hybridengine/engine.py [hunk 1]
```diff
@@ -8,31 +9,35 @@
...
-class PPInferEngine:
+_supported_models = [
+    "LlamaForCausalLM",
+]
+
+
+class CaiInferEngine:
     """
-    PPInferEngine is a class that handles the pipeline parallel inference.
+    InferEngine is a class that handles the pipeline parallel inference.
```
Comment (yuanheng-zhao) [inline colossalai/inference/hybridengine/engine.py:24]: Make the name consistent (`CaiInferEngine` or `InferEngine`)

File: colossalai/inference/hybridengine/engine.py [hunk 2]
```diff
@@ -42,7 +47,7 @@ class PPInferEngine:
     model = LlamaForCausalLM.from_pretrained("your_path_to_model")
     tokenizer = LlamaTokenizer.from_pretrained("/home/lczyh/share/models/llama-7b-hf")
     # assume the model is infered with 2 pipeline stages
-    inferengine = PPInferEngine(pp_size=2, model=model, model_policy=LlamaModelInferPolicy(), new_length=8)
+    inferengine = InferEngine(pp_size=2, model=model, model_policy=LlamaModelInferPolicy())
```
Comment (yuanheng-zhao) [inline colossalai/inference/hybridengine/engine.py:50]: preferably `infer_engine`

File: colossalai/inference/hybridengine/modeling/llama.py [hunk 3]
```diff
@@ -197,26 +256,23 @@ def llama_model_forward(
...
                 infer_state.decode_mem_start = alloc_mem[1]
                 infer_state.decode_mem_end = alloc_mem[2]
-                infer_state.block_loc[:, seq_length_with_past - 1] = infer_state.decode_mem_index
+                infer_state.block_loc[:, infer_state.max_len_in_batch - 1] = infer_state.decode_mem_index
             else:
                 print(f" *** Encountered allocation non-contiguous")
-                print(
-                    f"    infer_state.cache_manager.past_key_values_length: {infer_state.cache_manager.past_key_values_length}"
-                )
+                print(f"    infer_state.max_len_in_batch : {infer_state.max_len_in_batch}")
```
Comment (yuanheng-zhao) [inline colossalai/inference/hybridengine/modeling/llama.py:262]: remove prints for testing

File: colossalai/inference/hybridengine/engine.py (hunk 1 above)
Comment (FoolPlayer) [reply colossalai/inference/hybridengine/engine.py:24]: yes sir!

File: colossalai/inference/hybridengine/engine.py (hunk 2 above)
Comment (FoolPlayer) [reply colossalai/inference/hybridengine/engine.py:50]: got it!

File: colossalai/inference/hybridengine/modeling/llama.py (hunk 3 above)
Comment (FoolPlayer) [reply colossalai/inference/hybridengine/modeling/llama.py:262]: removed

**Review by yuanheng-zhao — APPROVED** [review]
LGTM, partially approved!

File: colossalai/inference/hybridengine/engine.py [hunk 4]
```diff
@@ -71,12 +75,19 @@ def __init__(
...
+        assert model and model_policy, "Model with model_policy should be provided."
         assert dtype in ["fp16", "fp32", "bf16"], "dtype should be one of 'fp16', 'fp32', 'bf16'"
 
-        max_output_len = max(max_output_len, max_input_len + new_length)
+        assert max_batch_size <= 64, "Max batch size exceeds the constraint"
+        assert max_input_len + max_output_len <= 4096, "Max length exceeds the constraint"
 
         self.pp_size = pp_size
+        self.tp_size = tp_size
+
```
Comment (tiandiao123) [inline colossalai/inference/hybridengine/engine.py:90]: can pipeline and tp-engine use together? If not, probably added a checking function to check: if pp_size > 1, assert tp_size == 1, and if tp_size > 1, assert pp_size != 1

Comment (FoolPlayer) [reply colossalai/inference/hybridengine/engine.py:92]: ok

**Review by tiandiao123 — APPROVED** [review]

**Pull request merged**